File: crud-framework-utils/src/main/java/com/antelopesystem/crudframework/utils/component/componentmap/ComponentMapPostProcessor.java
package com.antelopesystem.crudframework.utils.component.componentmap;

import com.antelopesystem.crudframework.utils.component.componentmap.annotation.ComponentMap;
import com.antelopesystem.crudframework.utils.component.componentmap.annotation.ComponentMapKey;
import com.antelopesystem.crudframework.utils.utils.ReflectionUtils;
import org.jetbrains.annotations.NotNull;
import org.springframework.aop.TargetClassAware;
import org.springframework.aop.framework.Advised;
import org.springframework.beans.BeansException;
import org.springframework.beans.factory.config.BeanPostProcessor;
import org.springframework.beans.factory.config.DependencyDescriptor;
import org.springframework.context.ApplicationContext;
import org.springframework.context.ApplicationContextAware;
import org.springframework.core.ResolvableType;
import org.springframework.core.annotation.AnnotationAwareOrderComparator;
import org.springframework.core.annotation.AnnotationUtils;

import java.lang.annotation.Annotation;
import java.lang.reflect.*;
import java.util.*;

public class ComponentMapPostProcessor implements BeanPostProcessor, ApplicationContextAware {

	private ApplicationContext applicationContext;

	@Override
	public void setApplicationContext(@NotNull ApplicationContext applicationContext) throws BeansException {
		this.applicationContext = applicationContext;
	}

	@Override
	public Object postProcessBeforeInitialization(@NotNull Object bean, @NotNull String beanName) throws BeansException {
		Object handler = bean;
		if(handler instanceof TargetClassAware) {
			try {
				handler = ((Advised) handler).getTargetSource().getTarget();
			} catch (Exception e) {
			}
		}

		List<Field> fields = ReflectionUtils.getFields(handler.getClass());

		for(Field field : fields) {
			if(field.isAnnotationPresent(ComponentMap.class) && field.getType() == Map.class) {

				ComponentMap mapped = field.getAnnotation(ComponentMap.class);
				try {
					DependencyDescriptor descriptor = new DependencyDescriptor(field, false);
					ResolvableType resolvableType = descriptor.getResolvableType().asMap();
					Class<?> keyClazz = resolvableType.resolveGeneric(0);
					Class<?> valueClazz = resolvableType.resolveGeneric(1);

					ReflectionUtils.makeAccessible(field);
					field.set(handler, initMap(keyClazz, valueClazz));
				} catch(Exception e) {
				}
			}
		}
		return bean;
	}

	private <T, E> Map<T, E> initMap(Class<T> key, Class<E> value) {
		Map<T, E> map = new LinkedHashMap<>();

		List<E> beans = new ArrayList<>(applicationContext.getBeansOfType(value).values());
		AnnotationAwareOrderComparator.sort(beans);
		for(E bean : beans) {
			List<Method> methods = ReflectionUtils.getMethods(bean.getClass());

			for(Method method : methods) {
				Annotation annotation = AnnotationUtils.findAnnotation(method, ComponentMapKey.class);
				if(annotation != null) {
					try {
						map.put((T) method.invoke(bean), bean);
					} catch(Exception e) {
					}
				}
			}
		}

		return map;
	}
}

File: crud-framework-utils/src/main/java/com/antelopesystem/crudframework/utils/component/componentmap/ComponentMapPostProcessor.kt
package com.antelopesystem.crudframework.utils.component.componentmap

import com.antelopesystem.crudframework.utils.component.componentmap.annotation.ComponentMap
import com.antelopesystem.crudframework.utils.component.componentmap.annotation.ComponentMapKey
import com.antelopesystem.crudframework.utils.component.componentmap.model.SingletonComponentMap
import com.antelopesystem.crudframework.utils.utils.ParameterLock
import com.antelopesystem.crudframework.utils.utils.ReflectionUtils
import com.antelopesystem.crudframework.utils.utils.getGenericClass
import com.antelopesystem.crudframework.utils.utils.resolveNestedGeneric
import org.apache.commons.lang3.ClassUtils
import org.springframework.aop.TargetClassAware
import org.springframework.aop.framework.Advised
import org.springframework.beans.factory.config.BeanPostProcessor
import org.springframework.core.annotation.AnnotationUtils
import java.lang.reflect.Method

class ComponentMapPostProcessor : BeanPostProcessor {

    private val componentMaps: MutableMap<ComponentMapIdentifier, MutableMap<Any, MutableList<Any>>> = mutableMapOf()

    override fun postProcessAfterInitialization(bean: Any, beanName: String?): Any {
        try {
            registerComponentMapKeyIfExists(bean)
            fillComponentMapIfExists(bean)
        } catch (e: Exception) {
        }
        return bean
    }

    private fun fillComponentMapIfExists(bean: Any) {
        var handler = bean
        if (handler is TargetClassAware) {
            try {
                handler = (handler as Advised).targetSource.target
            } catch (e: java.lang.Exception) {
            }
        }
        val fields = ReflectionUtils.getFields(handler.javaClass)

        for (field in fields) {
            if (field.isAnnotationPresent(ComponentMap::class.java)) {
                if (!Map::class.java.isAssignableFrom(field.type)) {
                    error("@ComponentMap may only be used on maps")
                }

                val mapped = field.getAnnotation(ComponentMap::class.java)
                try {
                    val keyClazz = field.getGenericClass(0)!!
                    var valueClazz = field.getGenericClass(1)!!
                    var isList = false
                    if(Collection::class.java.isAssignableFrom(valueClazz)) {
                        isList = true
                        valueClazz = field.resolveNestedGeneric(1)
                    }

                    ReflectionUtils.makeAccessible(field)
                    val map = getOrCreateComponentMap(keyClazz, valueClazz)
                    if(isList) {
                        field[handler] = map
                    } else {
                        field[handler] = SingletonComponentMap(map)
                    }
                } catch (e: java.lang.Exception) {
                }
            }
        }
    }

    private fun getOrCreateComponentMap(initialKeyType: Class<*>, initialValueType: Class<*>) : MutableMap<Any, MutableList<Any>> {
        val identifier = ComponentMapIdentifier(initialKeyType, initialValueType)
        var map = componentMaps[identifier]
        if(map != null) {
            return map
        }

        val lock = ParameterLock.getCanonicalParameterLock("${initialKeyType.canonicalName}_${initialValueType.canonicalName}")
        lock.lock()
        try {
            map = componentMaps[identifier]
            if(map != null) {
                return map
            }

            componentMaps[identifier] = mutableMapOf()
            return componentMaps[identifier]!!
        } finally {
            lock.unlock()
        }
    }

    private fun registerComponentMapKeyIfExists(bean: Any) {
        val methods = ReflectionUtils.getMethods(bean.javaClass)

        for (method in methods) {
            val annotation = AnnotationUtils.findAnnotation(method, ComponentMapKey::class.java)
            if (annotation != null) {
                try {
                    val key = method.invoke(bean)
                    val keyClass = key::class.java
                    val valueClass = getMethodDeclarer(method)
                    val map = getOrCreateComponentMap(keyClass, valueClass)
                    val list = map.computeIfAbsent(key) { mutableListOf() }
                    list += bean
                } catch (e: Exception) {
                }
            }
        }
    }


}

private data class ComponentMapIdentifier(val keyClazz: Class<*>, val valueClazz: Class<*>)

fun getMethodDeclarer(method: Method): Class<*> {
    var declaringClass: Class<*> = method.declaringClass
    val methodName: String = method.name
    val parameterTypes: Array<Class<*>> = method.parameterTypes
    for (interfaceType in ClassUtils.getAllInterfaces(declaringClass)) try {
        return interfaceType.getMethod(methodName, *parameterTypes).declaringClass
    } catch (ex: NoSuchMethodException) {
    }
    while (true) {
        declaringClass = declaringClass.superclass
        if (declaringClass == null) break
        try {
            val newMethod = declaringClass.getMethod(methodName, *parameterTypes)
            return getMethodDeclarer(newMethod)
        } catch (ex: NoSuchMethodException) {
            break
        }
    }

    error("Could not find method declarer for ${method.declaringClass.canonicalName}::${method.name}")
}

